syntax = "proto2";

package kangaroo.model;

message Account {
  enum Type {
    ROOT = 0;      // The parent of all accounts. Can only be defined once.

    ASSET = 1;         // Generic asset
    LIABILITY = 2;     // Generic liability
    EQUITY = 3;        // Equity
    INCOME = 4;        // Income
    EXPENSE = 5;       // Expense
    TRADING = 6;       // For currency/security exchanges
    CASH = 10;         // Cash

    CHECKING = 11;     // Checking account
    INVESTMENT = 12;   // Investment (stock, bond, etf, etc.)
    DEPOSIT = 13;      // Security deposit
    PREPAIDCARD = 14;  // Prepaid cards, gift cards

    SAVINGS = 18;      // Savings account
    BROKERAGE = 19;    // Brokerage account

    CREDITCARD = 20;

    PROPERTY = 30;

    ACC_PAY = 40;
    ACC_REC = 41;
  }

  // [Required] Unique identifier.
  optional int64 id = 1;

  // [Required] Type of account.
  optional Type type = 2;

  // [Required] User-defined name for the account.
  optional string name = 3;

  // [Optional] User-defined account idendifier.
  optional string code = 4;

  // [Required*] Parent of this account. *Except for the single root account.
  optional int64 parent_id = 5;

  // [Required] Transactions in closed accounts may not be changed.
  optional bool is_archived = 6 [default = true];

  // [Required] Placeholder accounts may not have transactions.
  optional bool is_placeholder = 7 [default = false];

  // [Required] Account commodity. Must be a security for Investment accounts, a currency otherwise.
  optional int64 commodity_id = 8;

  // [Optional]
  optional int64 institution_id = 9;

  // [Optional]
  optional int64 icon_id = 10;

  // [Optional]
  optional string note = 11;
}
